iroh/src/commands/provide.rs

```diff
@@ -0,0 +1,231 @@
+    let db = {
+        if iroh_data_root.is_dir() && marker.exists() {
+            // try to load db
+            Database::load(&iroh_data_root).await.with_context(|| {
+                format!(
+                    "Failed to load iroh database from {}",
+                    iroh_data_root.display()
+                )
+            })?
+        } else {
+            // directory does not exist, create an empty db
+            Database::default()
+        }
+    };
+    let key = Some(iroh_data_root.join("keypair"));
+
+    let provider = provide(
+        db.clone(),
+        addr,
+        key,
+        keylog,
+        rpc_port.into(),
+        config.derp_map(),
+        rt,
+    )
+    .await?;
+    let controller = provider.controller();
+
+    // task that will add data to the provider, either from a file or from stdin
+    let fut = {
+        let provider = provider.clone();
+        tokio::spawn(async move {
+            let (path, tmp_path) = if let Some(path) = path {
+                let absolute = path.canonicalize()?;
+                println!("Adding {} as {}...", path.display(), absolute.display());
+                (absolute, None)
+            } else {
+                // Store STDIN content into a temporary file
+                let (file, path) = tempfile::NamedTempFile::new()?.into_parts();
+                let mut file = tokio::fs::File::from_std(file);
+                let path_buf = path.to_path_buf();
+                // Copy from stdin to the file, until EOF
+                tokio::io::copy(&mut tokio::io::stdin(), &mut file).await?;
+                println!("Adding from stdin...");
+                // return the TempPath to keep it alive
+                (path_buf, Some(path))
+            };
+            // tell the provider to add the data
+            let stream = controller.server_streaming(ProvideRequest { path }).await?;
+            let (hash, entries) = aggregate_add_response(stream).await?;
+            print_add_response(hash, entries);
+            let ticket = provider.ticket(hash).await?;
+            println!("All-in-one ticket: {ticket}");
+            anyhow::Ok(tmp_path)
+        })
+    };
+
+    let provider2 = provider.clone();
+    tokio::select! {
+        biased;
+        _ = tokio::signal::ctrl_c() => {
+            println!("Shutting down provider...");
+            provider2.shutdown();
+        }
+        res = provider => {
+            res?;
+        }
+    }
+    // persist the db to disk.
+    db.save(&iroh_data_root).await?;
+
+    // the future holds a reference to the temp file, so we need to
+    // keep it for as long as the provider is running. The drop(fut)
+    // makes this explicit.
+    fut.abort();
+    drop(fut);
+    Ok(())
+}
+
+async fn provide(
+    db: Database,
+    addr: SocketAddr,
+    key: Option<PathBuf>,
+    keylog: bool,
+    rpc_port: Option<u16>,
+    dm: Option<DerpMap>,
+    rt: &runtime::Handle,
+) -> Result<Node> {
```

Comment:
possible to move this up into a new `iroh/src/provide.rs` file?

Reply:
This is very CLI specific, with connecting to RPC etc, so in this form that seems the wrong place. What functionality are you looking for? the core functionality is this

```
    let keypair = get_keypair(key).await?;

    let mut builder = Node::builder(db).keylog(keylog);
    if let Some(dm) = dm {
        builder = builder.derp_map(dm);
    }
    let builder = builder.bind_addr(addr).runtime(rt);
    builder.keypair(keypair).spawn().await?
```
which is already the api that was designed for external consumption.

Reply:
ah ok, I'll experiment with that.